server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/PerRecipientRateLimitMailet.scala

```diff
@@ -0,0 +1,159 @@
+  override def asString(): String = s"${
+    keyPrefix.map(prefix => prefix.value + "_")
+      .getOrElse("")
+  }${entityType.asString()}_${mailAddress.asString()}"
+}
+
+class PerRecipientRateLimitMailet @Inject()(rateLimiterFactoryProvider: RateLimiterFactoryProvider) extends GenericMailet {
+  private var exceededProcessor: String = _
+  private var rateLimiters: Seq[PerRecipientRateLimiter] = _
+
+  override def init(): Unit = {
+    val rateLimiterFactory: RateLimiterFactory = rateLimiterFactoryProvider.create(getMailetConfig)
+    val duration: Duration = parseDuration()
+    val keyPrefix: Option[KeyPrefix] = Option(getInitParameter("keyPrefix")).map(KeyPrefix)
+    exceededProcessor = getInitParameter("exceededProcessor", Mail.ERROR)
+
+    val countRateLimiter: Option[PerRecipientRateLimiter] = Option(getInitParameter("count"))
+      .flatMap(parameter => extractCountToRateLimiter(parameter, duration, rateLimiterFactory)
+        .map(rateLimiter => PerRecipientRateLimiter(rateLimiter, keyPrefix, Count)))
+
+    val sizeRateLimiter: Option[PerRecipientRateLimiter] = Option(getInitParameter("size"))
+      .map {
+        case "" => throw new IllegalArgumentException("'size' field cannot be empty if specified")
+        case s => s
+      }
+      .flatMap(parameter => extractSizeToRateLimiter(parameter, duration, rateLimiterFactory)
+        .map(rateLimiter => PerRecipientRateLimiter(rateLimiter, keyPrefix, Size)))
+
+    rateLimiters = Seq(countRateLimiter, sizeRateLimiter)
+      .filter(limiter => limiter.isDefined)
+      .map(limiter => limiter.get)
+  }
+
+  override def service(mail: Mail): Unit = {
+    if (!mail.getRecipients.isEmpty) {
+      val rateLimitResults: Seq[(MailAddress, RateLimitingResult)] = mail.getRecipients.asScala.toSeq
+        .map(recipient => (recipient, applyRateLimiter(mail, recipient)))
+
+      val rateLimitedRecipients: Seq[MailAddress] = rateLimitResults.filter(_._2.equals(RateExceeded)).map(_._1)
+      val acceptableRecipients: Seq[MailAddress] = rateLimitResults.filter(_._2.equals(AcceptableRate)).map(_._1)
+
+      // if all recipients exceed the rate limit -> overwrite the state mail property
+      // else cloning the mail (with exceeded recipients) and send the clone to the exceededProcessor
+      if (rateLimitedRecipients.nonEmpty) {
+        if (acceptableRecipients.isEmpty) {
+          mail.setState(exceededProcessor)
+        } else {
+          mail.setState(Mail.GHOST)
+          exceededProcess(mail, rateLimitedRecipients)
+        }
+      }
+
+      // if all recipients acceptable the rate limit -> do nothing
+      // else cloning the mail (with acceptable recipients) and send the clone to the next processor
+      if (acceptableRecipients.nonEmpty) {
+        if (rateLimitedRecipients.nonEmpty) {
+          mail.setState(Mail.GHOST)
+          acceptableProcess(mail, acceptableRecipients)
+        }
+      }
+    }
```

Comment:
to be honnest i do not like the way we specify the logic with double intricated if. pattern matching woul definitly be clearer.

```
(acceptableRecipients, rateLimitedRecipients) => {
  case (acceptable, exceeded) if acceptable.isEmpty => mail.setState(exceededProcessor)
  case (acceptable, exceeded) if exceeded.isEmpty => // do nothing
  case _ =>
     mail.setRecipients(acceptable)
     val newMail = mail.duplicate()
     newMail.setRecipients(exceeded)
     Using.resource(newMail)(getMailetContext.send(_, exceededProcessor)(LifecycleUtil.dispose)
}